pyabc/epsilon/temperature.py

```diff
@@ -0,0 +1,502 @@
+    observed to give big feasible temperature leaps in early iterations.
+
+    Parameters
+    ----------
+    target_rate: float
+        The target acceptance rate to match.
+    """
+
+    def __init__(self, target_rate: float = 0.5):
+        self.target_rate = target_rate
+
+    def __call__(self,
+                 t: int,
+                 get_weighted_distances: Callable,
+                 max_nr_populations: int,
+                 pdf_norm: float,
+                 kernel_scale: str,
+                 prev_temperature: float,
+                 acceptance_rate: float):
+        # execute function (expensive if in calibration)
+        df = get_weighted_distances()
+
+        weights = np.array(df['w'], dtype=float)
+        pdfs = np.array(df['distance'], dtype=float)
+
+        # compute rescaled posterior densities
+        if kernel_scale == SCALE_LIN:
+            values = pdfs / pdf_norm
+        else:  # kernel_scale == SCALE_LOG
+            values = np.exp(pdfs - pdf_norm)
+
+        # to acceptance probabilities
+        values = np.minimum(values, 1.0)
+
+        # to pmf
+        # TODO: We currently use the weights from the previous iteration.
+        # It might be better to use weights based on `transition / prior`
+        # for the next proposal transition kernel.
+        weights /= np.sum(weights)
+
+        # objective function which we wish to find a root for
+        def obj(beta):
+            val = np.sum(weights * values**beta) - self.target_rate
+            return val
+
+        if obj(1) > 0:
+            # function is mon. dec., smallest possible value already > 0
+            beta_opt = 1.0
+            # it is obj(0) >= 0 always
+        else:
+            # perform binary search
+            # TODO: check out more efficient optimization approach
+            beta_opt = sp.optimize.bisect(obj, 0, 1)
+
+        # temperature is inverse beta
+        temperature = 1.0 / beta_opt
+        return temperature
+
+
+class ExponentialDecayScheme(TemperatureScheme):
+    """
+    If `max_nr_populations` is finite:
+
+    .. math::
+        T_j = T_{max}^{(n-j)/n}
+
+    where n denotes the number of populations, and j=1,...,n the iteration.
+    This translates to
+
+    .. math::
+        T_j = T_{j-1}^{(n-j)/(n-(j-1))}.
+
+    This ensures that a temperature of 1.0 is reached after exactly the
+    remaining number of steps.
+
+    If `max_nr_populations` is infinite, the next temperature is
+
+    .. math::
+        T_j = \\alpha \\cdot T_{j-1},
+
+    where by default alpha=0.25.
+
+    So, in both cases the sequence of temperatures follows an exponential
+    decay, also known as a geometric progression, or a linear progression
+    in log-space.
+
+    Note that the formula is applied anew in each iteration. That has the
+    advantage that, if also other schemes are used s.t. T_{j-1} is smaller
+    than by the above, advantage can be made of this.
```

Comment:
"That has the advantage that, ... , advantage can be made of this." This sounds funny

Reply:
true. will reformulate.